### DISCUSSION TITLE
Merge All Primitive Logic/Math Elements in Circuit Boards

### DISCUSSION THREAD

**Discussion by Sviat** [FakeFishGames/Barotrauma#4818] [Ideas and suggestions]
### Problem
In-game UI overloaded with identical wiring components. Both Fabricator and Shop UI/UX suffers from this.
When some of them are hard to abstract and deserve separate entry (Wi-Fi, Color), biggest part of them is just proxy over simple math library functions.

### Solution
Merge Math elements (discrete) into two circuit boards: Unary Operator and Binary Operator.
Signal (continuous) units should be derived from Operators, but include additional OVERRIDE input signal.
Exact function should be selected via drop-down and lead to change of sprite. 

### Result
According to https://barotrauma.gamepedia.com/Wiring_Components before this change menu contains:
1. 8 logic units
2. 12 math operators
3. 6 trigonometry operators
4. 6 various uncategorized items
Total: 32 items - bad.

After change:
1. 1 Unary Signal Operator
2. 1 Binary Signal Operator
3. 1 Unary Operator
4. 1 Binary Operator
5. 8 various uncategorized items (added RegEx and Signal Check from Signals)
Total: 12 items - good.

**Comment by JamesTCrenshaw** [comment]
So, you have this object, and before or after you place the object you can set what function it is to do? (And subsequently change the sprite to reflect the function it fulfills?)

**Reply by Sviat** [reply]
After placement. 
In "wiring" view (above first Input and Output slots) you should be able to select exact function\operator at any moment or lock with lock to prevent accidental change.

**Comment by ubertpendragon** [comment]
Changing how all components work only to save UI space is insanity, but it's at least worth removing components from outpost shops since they're so easy to craft

**Reply by Sviat** [reply]
Insanity is to have one item per trigonometry or math operation.

Inventory, Fabricator and Shop are all suffering from this. Imagine new player opening craft panel for the first time. I expect 40% will just close it to never open again.

**Reply by ubertpendragon** [reply]
It's only a problem for the fabricator UI, where a good 90% of the components will never be used. 
All components can be removed from the shop UI, and it's not like you'll be holding a dozen components in your inventory.

Either you remove them from the shop, either you remove them from the fabricator, or the third option, having a separate "engineering" shop in stations with all the component junk. Medical has the same problem with the stupid amount of available syringes, but at least it has its own separate fabricator

**Reply by Sviat** [reply]
Med syringes at least all have different meaning, and frankly speaking - much more useful than El.components.
And there is difference (both in stats, resources and use-cases) even between opiates family, not talking difference between poisons and antidotes.

**Reply by Sviat** [reply]
>  and it's not like you'll be holding a dozen components in your inventory.

If you do Campaign and loot wrecked subs - you have 20-30 el.components incl. wires

**Comment by Working-Joe** [comment]
The way functionality of components are distrubuted across components is most likely to make wiring more accessible to newer players. Having a separate component for each separate function allows all sorts of players to do some sort of visual equation making similar to how e.g. visual scripting would make coding more accessible. 

Although combining them would make for less clutter in the crafting menu, it would cause greater inconvenience in making circuits. I'd rather see the components live in a circuit box like enclosure which can be crafted (see also #4721) than having all components combined into a single one that you have to set to a specific mode for each component.

It'd also be less accessible given that now all functionality of what would be different components is now hidden behind the interface of a combined component.

If your main gripe is the clutter in fabricators or crafting menus, then maybe a discussion considering UI changes there may be worthwhile too

**Reply by Sviat** [reply]
You probably playing on different continent.
I had 0 MP games where people with less than 100hours even wanted to touch electronics.

> combining them would make for less clutter in the crafting menu, it would cause greater inconvenience in making circuits. 

This is just not possible. It cannot be worse. Please make poll and distribute it via Discord \ Steam if you dont believe me.
Wiring mini-game is not for complete beginners.

**Reply by Working-Joe** [reply]
Just adding this response because I got this thread in my feed again because of your update :)

It Absolutely *would* be worse if each specific function of a component would be hidden behind a generic interface. Just a few reasons:

-You have to perform an extra step for each component (namely setting the primitive component to the right function)
-Players will have to learn in what primitive component each of the functions are. (Some comps have more or less inputs than you would expect or know)
-Components with different functions can look the same 

And that's besides the huge difficulty of implementing such a system in the first place. It'd be far more difficult to keep functionally different code separated than with the current system. Furthermore you'd now end up with odd component UI since the settings you have inside the component depend on the function you have chosen - I'm not even certain that is possible with the current itemcomponent system.

Lastly, the reason players don't want to touch wiring in game is because there's not much reason to and the wiring system is in itself rather convoluted. It does not necessarily have to do with the fact that there are so many different components. Categories in the fab UI would be a much better solution to the problem you pose, as proposed to you multiple times.

**Comment by JamesTCrenshaw** [comment]
I think this solution is far superior to other solutions, including UI changes. This change would solve a significant amount of inventory clutter issues, and would open the door to even further simplification down the line, should the devs have another similar issue with too many components that cannot stack due to their nature of being different functional items.

**Comment by Sviat** [comment]
Any word on this?

**Comment by mLuby** [comment]
Why not go all the way and replace _all_ the components with a real FPGA **Code Component**?

I think there are less drastic changes that could achieve the underlying aim, like a default-collapsed components submenu in vendor and fabricator lists, or being able to stack components of different types.

**Reply by Sviat** [reply]
if you thinks its less drastic i wish you goodluck stacking different kinds of components.

**Comment by leonard84** [comment]
I like this idea, and it would be great to have unary and binary versions of many components, e.g., greater than in unary would have one value configured on the component, instead of having to use a memory component as second input. There are several components, that already have this, e.g. signal check/regex. It would cut down on useless clutter for the case where you want to use a fixed value.

**Comment by Prolbo** [comment]
It's bad idea (to mess with existing logical elements). I didn't like it for many reasons.
Also this circuit board / micro controller concept is good, it'll be convenient to have programmable logical element with few ins and outs.
